src/relations/database_provides.py

```diff
@@ -83,22 +83,24 @@ def __init__(
     def _set_databag(
         self,
         *,
-        username: str,
-        password: str,
-        router_read_write_endpoint: str,
-        router_read_only_endpoint: str,
+        username: str = None,
+        password: str = None,
+        router_read_write_endpoint: str = None,
+        router_read_only_endpoint: str = None,
     ) -> None:
         """Share connection information with application charm."""
-        logger.debug(
-            f"Setting databag {self._id=} {self._database=}, {username=}, {router_read_write_endpoint=}, {router_read_only_endpoint=}"
-        )
-        self._interface.set_database(self._id, self._database)
-        self._interface.set_credentials(self._id, username, password)
-        self._interface.set_endpoints(self._id, router_read_write_endpoint)
-        self._interface.set_read_only_endpoints(self._id, router_read_only_endpoint)
-        logger.debug(
-            f"Set databag {self._id=} {self._database=}, {username=}, {router_read_write_endpoint=}, {router_read_only_endpoint=}"
-        )
+        if username and password:
```

Comment:
nit: I understand this to allow partial updates. why not just pass all values?

Reply:
when we update the endpoints (after the first time when the user/pass + endpoints are initially set), the only values changing are the endpoints (we do not have access to username or password). updating the username/password with None will wipe clean the values in the databag [here](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/database_provides.py#L231)